### DISCUSSION TITLE
Optional value with pattern

### DISCUSSION THREAD

**Discussion by careri** [json-schema-org/community#812] [Q&A] [answered]
Hi, I'm just getting started with schema validation. Nice concept, thanks!

I know I'm doing this wrong, but what would be the correct way of doing this?

I want a property **resource_url**
It can be valid in three different ways:
 1. **Null** is allowed
 2. **Empty** string is allowed
 3. A string that matches the **uri** format.

This is my attempt, it doesn't work:
```json
{
  "$id": "https://test.com/optional_value.json",
  "$schema": "https://json-schema.org/draft/2019-09/schema",
  "description": "Test",
  "properties": {
    "resource_url": {
      "description": "An optional string, should match the uri format if supplied"
    }
  },
  "type": "object",
  "anyOf": [
    {
      "if": {
        "properties": {
          "resource_url": {
            "type": "string",
            "minLength": 1
          }
        }
      },
      "then": {
        "properties": {
          "resource_url": {
            "format": "uri"
          }
        }
      }
    },
    {
      "if": {
        "properties": {
          "resource_url": {
            "type": "string",
            "maxLength": 0
          }
        }
      },
      "then": {
        "properties": {
          "resource_url": {
            "maxLength": 0
          }
        }
      }
    },
    {
      "if": {
        "properties": {
          "resource_url": {
            "type": "null"
          }
        }
      },
      "then": {
        "properties": {
          "resource_url": {
            "type": "null"
          }
        }
      }
    }
  ]
}
```

**Comment by jviotti** [comment, marked as answer]
Hi @careri , I think you can simplify your schema a lot by making use of `anyOf` instead of `if`/`then`/`else`. For example:

```json
{
  "$id": "https://test.com/optional_value.json",
  "$schema": "https://json-schema.org/draft/2019-09/schema",
  "description": "Test",
  "properties": {
    "resource_url": {
      "description": "An optional string, should match the uri format if supplied",
      "anyOf": [
        { "type": "null" },
        { "const": "" },
        { "type": "string", "minLength": 1, "format": "uri" },
      ]
    }
  }
}
```

Though keep in mind that not all JSON Schema implementations validate `format` by default (some only report it as metadata, which we call "annotations" in JSON Schema), so double check that you get the results you expect (as an aside, I built a unit testing framework for schemas: https://github.com/sourcemeta/jsonschema/blob/main/docs/test.markdown), and if not, you can still force-validate the URI directly using `pattern`.

**Reply by careri** [reply]
Hi, and thank you! That looks a lot clearer, I knew I was doing something wrong.

That worked perfectly :)
